### PR TITLE
fixes #22711; Check atomicArc for atomic destroy race condition

### LINKED ISSUE #22711: Check atomicArc for atomic destroy race condition
### Description

The `SharedPtr[T]` in [threading](https://github.com/nim-lang/threading/) had a race condition where sometimes no destructor would free the pointer. This occured when the atomic decrement isn't performed in the same atomic instruction as the counter check.

@Araq says this likely affects `atomicArc` as well. I tried to make a PR but don't have the time to understand the destructor codegen. 

For more details see: https://github.com/nim-lang/threading/issues/45 (note my first example was wrong! See the later one for the correct example)

### Nim Version

1.6.x and 2.0.x at least. 

### Current Output

_No response_

### Expected Output

_No response_

### Possible Solution

See https://github.com/nim-lang/threading/pull/46/ -- just change the `counter.load(Relexed)` to `counter.fetchSub(1, Release)`. The operation needs to be `ATOMIC_RELEASE` to force the proper ordering for the decrements. 

Also, the initial count may need to be changed depending if the `fetchSub` returns the original or the new count. 

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by ringabout**
fixes #22711

Per @elcritch's awesome solution

### CODE REVIEW

File: lib/system/arc.nim [hunk 1]
```diff
@@ -202,15 +202,22 @@ proc nimDecRefIsLast(p: pointer): bool {.compilerRtl, inl.} =
...
+      if atomicDec(cell.rc, rcIncrement) == -1:
+        result = true
+        when traceCollector:
+          cprintf("[ABOUT TO DESTROY] %p\n", cell)
     else:
-      decrement cell
-      # According to Lins it's correct to do nothing else here.
-      when traceCollector:
-        cprintf("[DECREF] %p\n", cell)
+      if cell.count == 0:
```
Comment (Araq) [inline lib/system/arc.nim:212]: Why? The old code was shorter and correct too.

Comment (ringabout) [reply lib/system/arc.nim:212]: What do you mean? I added a new branch for atomicArc and did nothing to the existing branch. It is said in the issue https://github.com/nim-lang/threading/issues/45 that they are the same problem.

Comment (ringabout) [reply lib/system/arc.nim:212]: The changed code is for `atomicArc`

It is changed from
```nim
proc `=destroy`*(x: ref) =
  if atomicLoad(x.rc) == 0:
    `=destroy`(x[])
  else:
    atomicDec(x.rc)
```

```nim
proc `=destroy`*(x: ref) =
  if atomicDec(x.rc) == -1:
    `=destroy`(x[])
```
according to the implementation in https://github.com/nim-lang/threading/pull/46

Comment (Araq) [reply lib/system/arc.nim:212]: My bad, I misread the diff.

**Pull request merged**